ibtk/src/lagrangian/FEDataManager.cpp

```diff
@@ -1399,21 +1422,34 @@ FEDataManager::interpWeighted(const int f_data_idx,
                     X_dof_map_cache.dof_indices(elem, X_dof_indices[d], d);
                 }
                 get_values_for_interpolation(X_node, *X_petsc_vec, X_local_soln, X_dof_indices);
-                const bool qrule_changed = updateInterpQuadratureRule(qrule, interp_spec, elem, X_node, patch_dx_min);
-                if (qrule_changed)
+                // TODO: like above, we only really need to know the number of
+                // quadrature points: we do not need to rebuild the quadrature
+                // rule.
+                updateInterpQuadratureRule(qrule, interp_spec, elem, X_node, patch_dx_min);
+                const std::pair<libMesh::QuadratureType, libMesh::Order> key(
+                    qrule->type(), qrule->get_order());
+                FEBase &F_fe = F_fe_cache[key];
+
+                // Like above: conditionally initialize the FE object if it is
+                // new
+                if (used_F_quadratures.find(key) == used_F_quadratures.end())
                 {
-                    // NOTE: Because we are only using the shape function values
-                    // for the FE object associated with X, we only need to
-                    // reinitialize X_fe whenever the quadrature rule changes.
-                    // In particular, notice that the shape function values
-                    // depend only on the element type and quadrature rule, not
-                    // on the element geometry.
-                    F_fe->attach_quadrature_rule(qrule.get());
-                    X_fe->attach_quadrature_rule(qrule.get());
-                    if (X_fe != F_fe) X_fe->reinit(elem);
+                    TBOX_ASSERT(F_fe_type.family == LAGRANGE);
+                    // get_phi changes the state of the object if called
+                    // before reinit
+                    F_fe.get_phi();
+                    F_fe.get_JxW();
+                    used_F_quadratures.insert(key);
                 }
-                F_fe->reinit(elem);
+
+                // Unlike above, we always reinit here since we need JxW
+                F_fe.reinit(elem);
+                const std::vector<std::vector<double> >& phi_F = F_fe.get_phi();
+                const std::vector<double> &JxW_F = F_fe.get_JxW();
```

Comment:
I'm not convinced that this is right: we don't really use this mapping, since we also have to take into account the deformation caused by `X`, which should alter the Jacobian (i.e., the contravariant transformation depends on `X`, right?)

Reply:
This routine is a little bit confusing, but what is happening is that we are setting the values for a RHS vector for an L2 projection, but we are not actually solving for the L2 projection. (That is what I mean when I say that it is a "weighted" interpolation.)

The projection is in reference coordinates. We would need to worry about `X` if we did this in current coordinates. Does this make sense?

Reply:
>  what is happening is that we are setting the values for a RHS vector for an L2 projection, but we are not actually solving for the L2 projection.

I was confused by that and did a lot of googling for 'weighted interpolations' before I figured out what was going on :) I'd vote to rename this to `setupLoadVector` or something like that (since the result is a dual space vector).

I think I am still confused about this: I probably just need to think harder to sort this out. We compute the quadrature points by reading `X` values, so the `J` in `JxW` needs to be computed w.r.t. that same coordinate transformation. Am I misunderstanding where we use the `X` data? It looks like we use it to compute the quadrature point locations in `get_values_for_interpolation`.

Reply:
We are sampling a field using the current coordinates of the quadrature points. `X` is the motion map from reference coordinates to current coordinates. The integration happens in reference coordinates, so `JxW` depends on the fixed shape of the element in the reference configuration, but is independent of the deformation from the reference configuration to the current configuration.

I think of this as happening in two stages. Using `s` as the mesh/reference coordinates and `X(s,t)` as the current (deformed) position of material point `s` at time `t`, first we evaluate `F_IB(s,t) = \sum_{i,j,k} f_{i,j,k} \delta_h(x_{i,j,k} - X(s,t)) h^3`. This is "the usual IB interpolation scheme".

We cannot use `F_IB` directly, however, because it is typically not directly representable in terms of the FE basis functions. So we define its projection `F` via `(F,phi_l)_2 = (F_IB,phi_l)_2 = \int(F_IB(s,t) phi_l(s) ds` for all FE basis functions `phi_l`. This is where we use `JxW` in this routine.